test/Microsoft.AspNetCore.Sockets.Common.Tests/Internal/Formatters/BinaryMessageFormatterTests.cs

```diff
@@ -0,0 +1,92 @@
+            foreach (var message in messages)
+            {
+                Assert.True(MessageFormatter.TryWriteMessage(message, output, MessageFormat.Binary));
+            }
+
+            Assert.Equal(expectedEncoding, output.ToArray());
+        }
+
+        [Theory]
+        [InlineData(new byte[] { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01 }, new byte[0])]
+        [InlineData(new byte[] { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04, 0x01, 0xAB, 0xCD, 0xEF, 0x12 }, new byte[] { 0xAB, 0xCD, 0xEF, 0x12 })]
+        public void WriteBinaryMessage(byte[] encoded, byte[] payload)
+        {
+            var message = MessageTestUtils.CreateMessage(payload);
+            var output = new ArrayOutput(chunkSize: 8); // Use small chunks to test Advance/Enlarge and partial payload writing
+
+            Assert.True(MessageFormatter.TryWriteMessage(message, output, MessageFormat.Binary));
+
+            Assert.Equal(encoded, output.ToArray());
+        }
+
+        [Theory]
+        [InlineData(new byte[] { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }, MessageType.Text, "")]
+        [InlineData(new byte[] { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x03, 0x00, 0x41, 0x42, 0x43 }, MessageType.Text, "ABC")]
+        [InlineData(new byte[] { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0B, 0x00, 0x41, 0x0A, 0x52, 0x0D, 0x43, 0x0D, 0x0A, 0x3B, 0x44, 0x45, 0x46 }, MessageType.Text, "A\nR\rC\r\n;DEF")]
+        [InlineData(new byte[] { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x03 }, MessageType.Close, "")]
+        [InlineData(new byte[] { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x11, 0x03, 0x43, 0x6F, 0x6E, 0x6E, 0x65, 0x63, 0x74, 0x69, 0x6F, 0x6E, 0x20, 0x43, 0x6C, 0x6F, 0x73, 0x65, 0x64 }, MessageType.Close, "Connection Closed")]
+        [InlineData(new byte[] { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02 }, MessageType.Error, "")]
+        [InlineData(new byte[] { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0C, 0x02, 0x53, 0x65, 0x72, 0x76, 0x65, 0x72, 0x20, 0x45, 0x72, 0x72, 0x6F, 0x72 }, MessageType.Error, "Server Error")]
+        public void WriteTextMessage(byte[] encoded, MessageType messageType, string payload)
+        {
+            var message = MessageTestUtils.CreateMessage(payload, messageType);
+            var output = new ArrayOutput(chunkSize: 8); // Use small chunks to test Advance/Enlarge and partial payload writing
+
+            Assert.True(MessageFormatter.TryWriteMessage(message, output, MessageFormat.Binary));
+
+            Assert.Equal(encoded, output.ToArray());
+        }
+
+        [Fact]
+        public void WriteInvalidMessages()
+        {
+            var message = new Message(new byte[0], MessageType.Binary, endOfMessage: false);
+            var output = new ArrayOutput(chunkSize: 8); // Use small chunks to test Advance/Enlarge and partial payload writing
```

Comment:
We should have a case where the size of the chunk is big enough to fit all data.